internal/encryption/keyprovider/aws_kms/provider.go

```diff
@@ -0,0 +1,98 @@
+package aws_kms
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+	"github.com/opentofu/opentofu/internal/encryption/keyprovider"
+)
+
+type keyMeta struct {
+	CiphertextBlob []byte `json:"ciphertext_blob"`
+}
+
+type keyProvider struct {
+	Config
+	svc *kms.Client
+	ctx context.Context
+}
+
+func (p keyProvider) Provide(rawMeta keyprovider.KeyMeta) (keyprovider.Output, keyprovider.KeyMeta, error) {
+	inMeta := rawMeta.(*keyMeta)
+	outMeta := keyMeta{}
+	out := keyprovider.Output{}
+
+	// Generate new key pair
+	if p.Symetric {
+		var spec types.DataKeySpec
+
+		for _, opt := range spec.Values() {
+			if string(opt) == p.KeySpec {
+				spec = opt
+			}
+		}
+
+		if len(spec) == 0 {
+			return out, outMeta, fmt.Errorf("Invalid key_spec %s, expected one of %v", p.KeySpec, spec.Values())
+		}
+
+		generatedKeyData, err := p.svc.GenerateDataKey(p.ctx, &kms.GenerateDataKeyInput{
+			KeyId:   aws.String(p.KMSKeyID),
+			KeySpec: spec,
+		})
+
+		if err != nil {
+			return out, outMeta, err
+		}
+
+		// Set inital outputs
+		out.EncryptionKey = generatedKeyData.Plaintext
+		out.DecryptionKey = generatedKeyData.Plaintext
+		outMeta.CiphertextBlob = generatedKeyData.CiphertextBlob
+	} else {
+		var spec types.DataKeyPairSpec
+
+		for _, opt := range spec.Values() {
+			if string(opt) == p.KeySpec {
+				spec = opt
+			}
+		}
+
+		if len(spec) == 0 {
+			return out, outMeta, fmt.Errorf("Invalid key_spec %s, expected one of %v", p.KeySpec, spec.Values())
+		}
+		generatedKeyData, err := p.svc.GenerateDataKeyPair(p.ctx, &kms.GenerateDataKeyPairInput{
+			KeyId:       aws.String(p.KMSKeyID),
+			KeyPairSpec: spec,
+		})
+
+		if err != nil {
+			return out, outMeta, err
+		}
+
+		// Set inital outputs
+		out.EncryptionKey = generatedKeyData.PublicKey
+		out.DecryptionKey = generatedKeyData.PrivateKeyPlaintext
```

Comment:
**If** we want to do assymetric crypto (and as indicated in the other comment, I'm against) then this is not the way to do it. Asymmetric keys must be used with an encryption method dedicated to them - RSA in this case most likely (not sure if KMS supports elliptic). RSA also has a very strong limit on the message size (I believe length of the key).

So the way this needs to be done is as follows:
- On encryption
  - Generate a new AES key.
  - Encrypt the AES key using the public key and put it into the metadata.
  - Put the encrypted private key into the metadata.
  - Pass the plaintext AES key out as the encryption key.
- On decryption
  - Decrypt the encrypted private key using kms.
  - Decrypt the AES key using the private key.
  - Pass the decrypted AES key as the decryption key.

But overall, as I said, I'd suggest we leave this out for now, and possibly add it in the future, if someone requests it.